gematria/datasets/annotating_importer.cc

```diff
@@ -0,0 +1,340 @@
+  if (!status.ok()) {
+    return status;
+  }
+
+  // Get the raw basic blocks, perf samples, and LBR data for annotation.
+  absl::StatusOr<std::vector<
+      std::tuple<uint64_t, uint64_t, std::vector<DisassembledInstruction>>>>
+      basic_blocks = GetBlocksFromELF(elf_file_name);
+  if (!basic_blocks.ok()) {
+    return basic_blocks.status();
+  }
+  const auto sample_types_and_samples = GetSamples();
+  if (!sample_types_and_samples.ok()) {
+    return sample_types_and_samples.status();
+  }
+  const auto &[sample_types, samples] = sample_types_and_samples.value();
+  const auto &lbr_data = GetLBRData();
+  if (!lbr_data.ok()) {
+    return lbr_data.status();
+  }
+
+  // Convert the raw basic blocks into protos and annotate them using samples.
+  std::vector<BasicBlockWithThroughputProto> basic_block_protos(
+      basic_blocks->size());
+  for (int block_idx = 0; block_idx < basic_blocks->size(); ++block_idx) {
+    // Create the un-annotated basic block proto.
+    const auto &[block_begin_addr, block_end_addr, instructions] =
+        basic_blocks->at(block_idx);
+    auto &basic_block_proto = basic_block_protos[block_idx];
+    *basic_block_proto.mutable_basic_block() =
+        importer_.BasicBlockProtoFromInstructions(instructions);
+
+    // Loop over and annotate individual instructions.
+    for (int instruction_idx = 0; instruction_idx < instructions.size();
+         ++instruction_idx) {
+      uint64_t instruction_addr = basic_block_proto.basic_block()
+                                      .machine_instructions()[instruction_idx]
+                                      .address();
+      if (!samples.count(instruction_addr)) {
+        continue;
+      }
+      const std::vector<int> &annotations = samples.at(instruction_addr);
+
+      auto &instruction_proto = basic_block_proto.mutable_basic_block()
+                                    ->mutable_canonicalized_instructions()
+                                    ->at(instruction_idx);
+      for (int annotation_idx = 0; annotation_idx < annotations.size();
+           ++annotation_idx) {
+        if (annotations[annotation_idx]) {
+          *instruction_proto.add_instruction_annotations() =
+              ProtoFromAnnotation(Annotation(
+                  /* name = */ sample_types.at(annotation_idx),
+                  /* value = */ annotations.at(annotation_idx)));
+        }
+      }
+    }
+
+    ThroughputWithSourceProto &throughput =
+        *basic_block_protos[block_idx].add_inverse_throughputs();
+    throughput.set_source(source_name);
+    double throughput_value = -1;
+    if (lbr_data.value().count(block_begin_addr)) {
+      const auto &[run_end_addr, cycles_values] =
+          lbr_data.value().at(block_begin_addr);
+      if (block_end_addr == run_end_addr) {
+        throughput_value =
+            std::accumulate(cycles_values.begin(), cycles_values.end(), 0.0) /
+            cycles_values.size();
+      }
+    }
```

Comment:
We might want to take the LBR (pseudo-)blocks. By going through BBAddrMap, we might miss a large number of blocks where we have LBR data, but they do not correspond to a BBAddrMap entry.

This doesn't concern just traces with a not-taken conditional branch, but also branchless instruction sequences that can be entered at multiple addresses:
```
L1:
  inst1
  inst2
L2:
  inst3
  inst4
L3:
  inst5
  inst6
  JMP Lxyz
```
will appear in BBAddrMap as L1-L2, L2-L3, L3-JMP, but we would get L1-JMP, L2-JMP, L3-JMP from LBR and we'd throw out L1-JMP, L2-JMP.